### DISCUSSION TITLE
Headless Rendering within a Docker Container

### DISCUSSION THREAD

**Discussion by smmislam** [google-deepmind/mujoco#1995] [Asking for Help]
Hi,

I'm a graduate student, using MuJoCo for training RL models. I want to run my models within a docker container.
I'm looking for some help with headless rendering.

First off, below is the docker-file that I'm using to create docker-image.

```
FROM python:3.8.10

RUN python -m pip install --upgrade pip

RUN pip install torch==2.1.1 torchvision==0.16.1 torchaudio==2.1.1 --index-url https://download.pytorch.org/whl/cu118

RUN python -m pip install dm_control gymnasium
RUN python -m pip install numpy matplotlib seaborn einops ml_collections 
RUN python -m pip install wandb moviepy imageio

RUN apt-get update
RUN apt-get install -y libglew2.1
RUN apt-get install libgl1-mesa-glx libosmesa6
```

My script tries to initialize the walker-walk environment and do some training. I have set-up the exact same environment using conda, and everything is fine and working. However, when I try to run the same script from within the docker container I get this error below

/usr/local/lib/python3.8/site-packages/glfw/__init__.py:914: GLFWError: (65537) b'The GLFW library is not initialized'
  warnings.warn(message, GLFWError)
.........
  File "/usr/local/lib/python3.8/site-packages/dm_control/suite/wrappers/pixels.py", line 77, in __init__
    pixels = env.physics.render(**render_kwargs)
  File "/usr/local/lib/python3.8/site-packages/dm_control/mujoco/engine.py", line 223, in render
    camera = Camera(
  File "/usr/local/lib/python3.8/site-packages/dm_control/mujoco/engine.py", line 711, in __init__
    if self._physics.contexts.mujoco is not None:
  File "/usr/local/lib/python3.8/site-packages/dm_control/mujoco/engine.py", line 533, in contexts
    self._make_rendering_contexts()
  File "/usr/local/lib/python3.8/site-packages/dm_control/mujoco/engine.py", line 519, in _make_rendering_contexts
    mujoco_context = wrapper.MjrContext(self.model, render_context)
  File "/usr/local/lib/python3.8/site-packages/dm_control/mujoco/wrapper/core.py", line 608, in __init__
    ptr = ctx.call(mujoco.MjrContext, model.ptr, font_scale)
  File "/usr/local/lib/python3.8/site-packages/dm_control/_render/executor/render_executor.py", line 138, in call
    return func(*args, **kwargs)
mujoco.FatalError: an OpenGL platform library has not been loaded into this process, this most likely means that a valid OpenGL context has not been created before mjr_makeContext was called

To solve this, I changed the environment variable like so
`export MUJOCO_GL=egl`

And finally, I get the following error
...
from dm_control import suite
  File "/usr/local/lib/python3.8/site-packages/dm_control/suite/__init__.py", line 24, in <module>
    from dm_control.suite import acrobot
  File "/usr/local/lib/python3.8/site-packages/dm_control/suite/acrobot.py", line 20, in <module>
    from dm_control import mujoco
  File "/usr/local/lib/python3.8/site-packages/dm_control/mujoco/__init__.py", line 18, in <module>
    from dm_control.mujoco.engine import action_spec
  File "/usr/local/lib/python3.8/site-packages/dm_control/mujoco/engine.py", line 41, in <module>
    from dm_control import _render
  File "/usr/local/lib/python3.8/site-packages/dm_control/_render/__init__.py", line 86, in <module>
    Renderer = import_func()
  File "/usr/local/lib/python3.8/site-packages/dm_control/_render/__init__.py", line 36, in _import_egl
    from dm_control._render.pyopengl.egl_renderer import EGLContext
  File "/usr/local/lib/python3.8/site-packages/dm_control/_render/pyopengl/egl_renderer.py", line 39, in <module>
    from dm_control._render.pyopengl import egl_ext as EGL
  File "/usr/local/lib/python3.8/site-packages/dm_control/_render/pyopengl/egl_ext.py", line 29, in <module>
    from OpenGL import EGL
  File "/usr/local/lib/python3.8/site-packages/OpenGL/EGL/__init__.py", line 2, in <module>
    from OpenGL.raw.EGL._types import *
  File "/usr/local/lib/python3.8/site-packages/OpenGL/raw/EGL/_types.py", line 87, in <module>
    raw_eglQueryString = _p.PLATFORM.EGL.eglQueryString
AttributeError: 'NoneType' object has no attribute 'eglQueryString'

I would really appreciate any help or suggestion on how to solve this problem.

Note that I was able to run the script with the X Virtual Framebuffer apprach suggested in [issue#350](https://github.com/google-deepmind/dm_control/issues/350). But, I suspect that the rendering doesn't use GPU and hence quite slow.

**Comment by AdityaPradhan12** [comment]
I am facing the same issue. Were you able to solve it?

**Comment by michaelala25** [comment]
Also experiencing the same issue. Managed to get it solved by running `xhost +local:docker` to give docker X11 server access, launching the docker container as:
```
docker run -it --env DISPLAY=$DISPLAY --volume /tmp/.X11-unix:/tmp/.X11-unix --device /dev/dri:/dev/dri <image-tag>
```
and then exporting `LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libstdc++.so.6` inside the container (you can find your file with `find / -name libstdc++.so.6 2>/dev/null`).

If you have the nvidia-container-toolkit installed it also _appears_ to sorta work if you replace `--device /dev/dri:/dev/dri` with `--gpus all`, but rendering seems to happen on the CPU rather than the GPU (and you'll get a bunch of `libGL` errors complaining that it failed to find some devices, failed to load some drivers, etc.).

Still unsure if there's a way to do this without explicit X11 forwarding. Ideally what I want is to be able to render using OpenGL, but not necessarily _display_ things to a window (e.g. so I can simulate a camera as input to a neural network).

**Reply by michaelala25** [reply]
This may also be relevant for rending with EGL in a headless environment: https://forums.developer.nvidia.com/t/unable-to-initialize-egl/242372/2

**Comment by Andrew-Luo1** [comment]
I built cudagl with ubuntu 22.04 and cuda 12.4.0. In this container, after running `export MUJOCO_GL=egl`, the headless rendering works, no X11 forwarding! The rendering is as fast in the container as on the host.

**Reply by smmislam** [reply]
Hey! Thank you for the response!

Would you mind sharing the docker commands that you used to create this image with me?

**Reply by Andrew-Luo1** [reply]
Hi, yes apologies for the lack of details. I had forgotten to complete my original response. I've written up a tutorial on what I did to get this working [here](https://github.com/Andrew-Luo1/Mujoco-Headless-Tutorial/tree/main). Hopefully it helps anyone who's interested in Headless rendering in Mujoco going forward!

**Comment by OliEfr** [comment]
I got mujoco headless GPU-accelerated rendering in to work in a docker container with the following `Dockerfile`. Its based on a [nvidia container](https://catalog.ngc.nvidia.com/containers). You need a nvidia gpu for this. You can check if your container has access to the gpu using [this Issue](https://github.com/mmatl/pyrender/issues/149).

```
FROM nvcr.io/nvidia/pytorch:24.10-py3

# install requirements for your python project
COPY requirements.txt .
RUN pip install --no-cache-dir -r requirements.txt

# for headless GPU rendering
ENV MUJOCO_GL="egl"
ENV PYOPENGL_PLATFORM="egl"

#requirements for headless GPU rendering, see also https://github.com/mmatl/pyrender/issues/149
ENV NVIDIA_DRIVER_CAPABILITIES=compute,graphics,utility,video
RUN apt-get update && apt-get install -y --no-install-recommends \
        libegl1\
        libgles2 && \
        rm -rf /var/lib/apt/lists/*
RUN echo '{"file_format_version": "1.0.0", "ICD": {"library_path": "libEGL_nvidia.so.0"}}' >> /usr/share/glvnd/egl_vendor.d/10_nvidia.json

# Project specific commands go here
...
```

**Comment by DmytroMelnyk** [comment]
If the machine where you want to host your MuJoCo app doesn't have a GPU, you can use the `osmesa` backend (software renderer) for `MUJOCO_GL`. My Dockerfile looks like this:

```Dockerfile
FROM python:3.13-slim
RUN apt-get update && apt-get install -y libosmesa6-dev
WORKDIR /app
COPY . /app
RUN pip install --no-cache-dir -r requirements.txt
ENV MUJOCO_GL=osmesa
```

**Comment by cooijmanstim** [comment]
I had the same problem, when trying to use the EGL backend for headless rendering, not inside a container but on a cloud instance. What solved it for me was installing `libegl1-mesa-dev` with apt-get.
